src/vs/editor/common/viewModel/viewModelImpl.ts

```diff
@@ -249,6 +250,78 @@ export class ViewModel extends Disposable implements IViewModel {
 	}
 
 	private _registerModelEvents(): void {
+		this._register(this.model.onDidChangeDecorations(e => {
+			if (!e.affectedInjectedTextLines || e.affectedInjectedTextLines.size === 0) {
```

Comment:
This event suffers from the same problem we talked about regarding event ordering. For example, it might be possible for another listener of `onDidChangeDecorations` to get invoked before this listener. That other listener could, while being invoked with the `onDidChangeDecorations` event, go back to the text model and issue an `applyEdit`. That will cause that the `onDidChangeRawContentFast` event arrives at the view model (see below) before this `onDidChangeDecorations` event. So it is possible to receive a stale `onDidChangeDecorations` event. So I suggest that we include a `versionId` field in `IModelDecorationsChangedEvent` and then decide what we do in the case where we get a stale `IModelDecorationsChangedEvent` event. We might simply drop everything and recreate the entire lines if that rare case occurs. Also, I think we should just include all the needed data in the event itself, so it might be good for the event to contain the line contents and the injected texts for all affected lines.

That is why I suggested to try to reuse `ModelRawContentChangedEvent`. I still think that would be the better way forward. I think the `DidChangeContentEmitter` class inside `textModel.ts` could be changed to accomodate this new use-case such that it is possible to emit a `ModelRawContentChangedEvent` in a more simple way. The event could come with a real `versionId` or `undefined` or a new property can be added to `ModelRawContentChangedEvent` to indicate that this is not a real change at all, it is just about injected text.